### ISSUE TITLE
gRPC subscription fails with UNAVAILABLE EOF Error

### ISSUE THREAD

**Issue by dinesh707** [grpc/grpc-java#3800]
Please answer these questions before submitting your issue.

### What version of gRPC are you using?
grpc stub 1.7

### What did you expect to see?
no failiure

1st I do a Login call using the same ManagedChannel, which works fine. But then the next subscription request always fails with similar error as following.

Following is the stack trace.
io.grpc.StatusRuntimeException: UNAVAILABLE: {"created":"@1511823153.366760371","description":"EOF","file":"../../../../../../../../src/dist/grpc/src/core/lib/iomgr/tcp_posix.c","file_line":235,"grpc_status":14}��

	at io.grpc.Status.asRuntimeException(Status.java:526) ~[grpc-core-1.7.0.jar:1.7.0]
	at io.grpc.stub.ClientCalls$StreamObserverToCallListenerAdapter.onClose(ClientCalls.java:385) ~[grpc-stub-1.7.0.jar:1.7.0]
	at io.grpc.ForwardingClientCallListener.onClose(ForwardingClientCallListener.java:41) ~[grpc-core-1.7.0.jar:1.7.0]
	at io.grpc.internal.CensusTracingModule$TracingClientInterceptor$1$1.onClose(CensusTracingModule.java:339) ~[grpc-core-1.7.0.jar:1.7.0]
	at io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:443) ~[grpc-core-1.7.0.jar:1.7.0]
	at io.grpc.internal.ClientCallImpl.access$300(ClientCallImpl.java:63) ~[grpc-core-1.7.0.jar:1.7.0]
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.close(ClientCallImpl.java:525) ~[grpc-core-1.7.0.jar:1.7.0]
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.access$600(ClientCallImpl.java:446) ~[grpc-core-1.7.0.jar:1.7.0]
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:557) ~[grpc-core-1.7.0.jar:1.7.0]
	at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37) ~[grpc-core-1.7.0.jar:1.7.0]
	at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:107) ~[grpc-core-1.7.0.jar:1.7.0]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_144]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_144]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_144]

**Comment by dinesh707** [comment]
It seems that I get some amount of data in each response, and it get killed while its reading.

**Comment by dinesh707** [comment]
Here is the explanation of the bug in detail 
https://stackoverflow.com/questions/47541037/grpc-unavailable-eof-error-while-waiting-for-data?noredirect=1#comment82043457_47541037

**Comment by dinesh707** [comment]
disabling tracing fixed the issue. (So it would be nice that this flag setting is available as.a public method)

     ```
Field declaredField = NettyChannelBuilder.class.getSuperclass().getDeclaredField("tracingEnabled");
declaredField.setAccessible(true);
declaredField.set(nettyChannelBuilder, false);
```

**Comment by carl-mastrangelo** [comment]
cc: @zhangkun83

**Comment by zhangkun83** [comment]
@dinesh707 how did you find out tracing was related?

**Comment by zhangkun83** [comment]
From the post on SO, `CensusTracingModule$TracingClientInterceptor` does show up in the stack trace. However, it's not an issue by itself, because any interceptor that intercepts `onClose()` would be there.

**Comment by dinesh707** [comment]
We did some TCP dumps and found out the difference between jtimon (go) application and our java code. The Java one had the header
"Header: grpc-trace-bin: (Name Length: 14, Name: grpc-trace-bin, Value Length: 0, Value: Representation: Indexed Header Field Index: 63)"

**Comment by zhangkun83** [comment]
The header doesn't look suspicious. I still can't associate the header with the error. The error message appears to be generated from C++ code. Is the server C++?

**Comment by dinesh707** [comment]
Its written by Juniper. 
https://www.juniper.net/documentation/en_US/junos/topics/concept/junos-telemetry-interface-oveview.html

But why does the grpc-java code sets headers which are not required?
And why the method for setting "tracingEnabled" in "protected"?

**Comment by zhangkun83** [comment]
Sounds to me that Juniper doesn't handle unexpected headers gracefully. It's not a bug of gRPC per se, but we can optimize out empty stats/tracing headers so that it won't annoy you.

**Comment by creamsoup** [comment]
Resolved by #4805
